jax/_src/nn/functions.py

```diff
@@ -843,11 +843,51 @@ def _apply_masks(logits, mask, is_causal, q_seqlen, kv_seqlen,
   padded_logits = jnp.where(combined_mask, logits, large_negative_number)
   return padded_logits
 
+# We use this custom dot_general in the QK einsum op of attention to match
+# dtypes used in the Flash Attention implementation. For bf16 inputs as an
+# example, the fprop is like:
+#   bf16 -> dot -> fp32
+# Then the bprop is like:
+# (1) Without this change:
+#   fp32 -> dot -> fp32 -> cvt -> bf16.
+# (2) With this change:
+#   fp32 -> cvt -> bf16 -> dot -> bf16.
+@partial(custom_jvp, nondiff_argnums=(2, 3, 4))
+def dot_general_with_custom_convert(
+    lhs, rhs, dimension_numbers, precision=None, preferred_element_type=None
+):
+  return lax.dot_general(
+      lhs, rhs, dimension_numbers, precision=precision,
+      preferred_element_type=preferred_element_type,
+  )
+
+@dot_general_with_custom_convert.defjvp
+def dot_general_with_custom_convert_jvp(
+    dimension_numbers, precision, preferred_element_type, primals, tangents
+):
+  lhs, rhs = primals
+  lhs_dot, rhs_dot = tangents
+
+  out = lax.dot_general(
+      lhs, rhs, dimension_numbers, precision=precision,
+      preferred_element_type=preferred_element_type,
+  )
+
+  grad_out = lax.dot_general(
+      lhs_dot, rhs, dimension_numbers, precision=precision,
+  ) + lax.dot_general(
+      lhs, rhs_dot, dimension_numbers, precision=precision,
+  )
+  grad_out = grad_out.astype(preferred_element_type)
```

Comment:
Can you talk through the argument here in a bit more detail? This doesn't seem like a great idea, because it's not clear to me why we would want to accumulate the tangents using a different dtype than the primals. I see why this ends up giving the correct dtypes on the backwards pass, but it seems bad for the numerics of fwd mode.

Perhaps there's something I'm missing here, but if not, we'd probably want to use `custom_vjp` instead of `custom_jvp` because then it's clear that the fwd pass isn't appropriate.

WDYT?

Reply:
The main reason we use jvp instead of vjp is that we don't need to compute the new `dimension_number` for the bprop dots of grad_x and grad_w.

And I agree the code is a bit hacky and may be that clear on its purpose. I will try your dot algorithm since it works now as you mentioned.

Will update the thread later. Thx.